File: z/二叉搜索树与双向链表.cc
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};*/
class Solution {
 public:
  void helper(TreeNode* root, TreeNode*& prev) {
    if (root == NULL) {
      return;
    }
    helper(root->left, prev);
    root->left = prev;
    if (prev != NULL) {
      prev->right = root;
    }
    prev = root;
    helper(root->right, prev);
  }
  TreeNode* Convert(TreeNode* pRootOfTree) {
    if (pRootOfTree == NULL) {
      return NULL;
    }
    TreeNode* prev = NULL;
    helper(pRootOfTree, prev);
    TreeNode* res = pRootOfTree;
    while (res->left != NULL) {
      res = res->left;
    }
    return res;
  }
};

File: h/二叉搜索树与双向链表.java
/*
输入一棵二叉搜索树，将该二叉搜索树转换成一个排序的双向链表。
要求不能创建任何新的结点，只能调整树中结点指针的指向。
*/
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public TreeNode Convert(TreeNode pRootOfTree) {
        TreeNode lastNode = null;
			TreeNode headNode = ConvertNode(pRootOfTree, lastNode);
			while (headNode != null && headNode.left != null) {
				headNode = headNode.left;
			}
			return headNode;
		}
 
		public static TreeNode ConvertNode(TreeNode rootTree, TreeNode lastNode) {
			if (rootTree == null) {
				return null;
			}
			if (rootTree.left != null) {
				lastNode = ConvertNode(rootTree.left, lastNode);
			}
			rootTree.left = lastNode;
			if (lastNode != null) {
				lastNode.right = rootTree;
			}
			lastNode = rootTree;
			if (rootTree.right != null) {
				lastNode=ConvertNode(rootTree.right, lastNode);
			}
			return lastNode;
    }
}
